viewing_party/party.py

```diff
@@ -1,23 +1,176 @@
+        rating_list.append(movie["rating"])
+    average_rating = sum(rating_list) / len(rating_list)
+    return average_rating
+
+
+def get_most_watched_genre(user_data):
+    if not user_data["watched"]:
+        return None
+    
+    genre_list = []
+    for movie in user_data["watched"]:
+        genre_list.append(movie["genre"])
+    
+    most_watched_genre_frequency = 0
+    most_watched_genre = genre_list[0]
+    for genre in genre_list:
+        genre_frequency = genre_list.count(genre)
+        if genre_frequency > most_watched_genre_frequency:
+            most_watched_genre_frequency = genre_frequency
+            most_watched_genre = genre
+
+    return most_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+
+    user_watched_titles =[]
+    for movie in user_data["watched"]:
+        user_watched_titles.append(movie["title"])
+    
+    friends_watched_titles = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in friends_watched_titles:
+                friends_watched_titles.append(movie["title"])
+    
+    user_unique_titles = []
+    for title in user_watched_titles:
+        if title not in friends_watched_titles:
+            user_unique_titles.append(title)
+    
+    user_unique_movies = []
+    for title in user_unique_titles:
+        for movie in user_data["watched"]:
+            if movie["title"] == title:
+                user_unique_movies.append(movie)
+
+    return user_unique_movies
+
+
+def get_friends_unique_watched(user_data):
+    
+    user_watched_titles =[]
+    for movie in user_data["watched"]:
+        user_watched_titles.append(movie["title"])
+    
+    friends_watched_titles = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in friends_watched_titles:
+                friends_watched_titles.append(movie["title"])
+    
+    friends_unique_titles = []
+    for title in friends_watched_titles:
+        if title not in user_watched_titles:
+            friends_unique_titles.append(title)
+    
+    friends_unique_movies = []
+    for title in friends_unique_titles:
+        for friend in user_data["friends"]:
+            for movie in friend["watched"]:
+                if movie["title"] == title and movie not in friends_unique_movies:
+                    friends_unique_movies.append(movie)
+
+    return friends_unique_movies
+
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+
+    friends_recommended_movies = get_friends_unique_watched(user_data)
+
+    recommended_movies =[]
+    for movie in friends_recommended_movies:
+        if movie["host"] in user_data["subscriptions"]:

```

Comment:
What's the time complexity of a for loop that has to execute the in operator on a list during every iteration?

We could improve the subscription lookup if we trade a little space and make a set of the subscriptions (which are strings, so can go in a set). Using in on a list is a linear operation, but on a set, it's constant.

```python
subscriptions = set(user_data["subscriptions"]) 
for movie in friends_recommended_movies: 
    if movie["host"] in subscriptions: 
        recommended_movies.append(movie)
```